lib/filterx/filterx-metrics.c

```diff
@@ -0,0 +1,241 @@
+}
+
+static const gchar *
+_format_sck_name(FilterXMetrics *self)
+{
+  if (self->key.str)
+    return self->key.str;
+
+  FilterXObject *key_obj = filterx_expr_eval_typed(self->key.expr);
+  if (!key_obj)
+    return NULL;
+
+  gsize len;
+  const gchar *name = filterx_string_get_value(key_obj, &len);
+  if (len == 0)
+    {
+      filterx_eval_push_error("failed to format metrics key: key must be a non-empty string", self->key.expr, key_obj);
+      goto exit;
+    }
+
+  GString *name_buffer = scratch_buffers_alloc();
+  g_string_append(name_buffer, name);
+  name = name_buffer->str;
+
+exit:
+  filterx_object_unref(key_obj);
+  return name;
+}
+
+static gboolean
+_format_sck(FilterXMetrics *self, StatsClusterKey *sck)
+{
+  const gchar *name = _format_sck_name(self);
+  if (!name)
+    return FALSE;
+
+  StatsClusterLabel *labels;
+  gsize labels_len;
+  if (!filterx_metrics_labels_format(self->labels, &labels, &labels_len))
+    return FALSE;
+
+  stats_cluster_single_key_set(sck, name, labels, labels_len);
+  return TRUE;
+}
+
+static gboolean
+_is_const(FilterXMetrics *self)
+{
+  return !self->key.expr && (!self->labels || filterx_metrics_labels_is_const(self->labels));
+}
+
+gboolean
+filterx_metrics_get_stats_counter(FilterXMetrics *self, StatsCounterItem **counter)
+{
+  if (!filterx_metrics_is_enabled(self))
+    {
+      *counter = NULL;
+      return TRUE;
+    }
+
+  if (_is_const(self))
+    {
+      *counter = stats_cluster_single_get_counter(self->const_cluster);
+      return TRUE;
+    }
+
+  gboolean success = FALSE;
+
+  ScratchBuffersMarker marker;
+  scratch_buffers_mark(&marker);
+
+  StatsClusterKey sck;
+  if (!_format_sck(self, &sck))
+    goto exit;
+
+  *counter = metrics_cache_get_counter(metrics_tls_cache(), &sck, self->level);
+  success = TRUE;
+
+exit:
+  scratch_buffers_reclaim_marked(marker);
+  return success;
+}
+
+void
+filterx_metrics_free(FilterXMetrics *self)
+{
+  filterx_expr_unref(self->key.expr);
+  g_free(self->key.str);
+
+  if (self->labels)
+    filterx_metrics_labels_free(self->labels);
+
+  stats_lock();
+  {
+    StatsCounterItem *counter = stats_cluster_single_get_counter(self->const_cluster);
+    stats_unregister_dynamic_counter(self->const_cluster, SC_TYPE_SINGLE_VALUE, &counter);
+  }
+  stats_unlock();
+
+  g_free(self);
+}
+
+static gboolean
+_init_key(FilterXMetrics *self, FilterXExpr *key)
+{
+  if (!filterx_expr_is_literal(key))
+    {
+      self->key.expr = filterx_expr_ref(key);
+      return TRUE;
+    }
+
+  FilterXObject *key_obj = filterx_expr_eval_typed(key);
+  if (!filterx_object_is_type(key_obj, &FILTERX_TYPE_NAME(string)))
+    {
+      filterx_eval_push_error_info("failed to init metrics key, key must be a string", key,
+                                   g_strdup_printf("got %s instread", key_obj->type->name), TRUE);
+      filterx_object_unref(key_obj);
+      return FALSE;
+    }
+
+  self->key.str = g_strdup(filterx_string_get_value(key_obj, NULL));
```

Comment:
probably filterx_object_extract_string() would make sense here (depends on PR #229)

Reply:
maybe the plan is to merge this first, in that case of course #229 should address this.

Reply:
In this specific scenario, we don't need to call the extract function, as we check before that it is a literal, and there are no literal message values.